test/widgets/test_clock.py

```diff
@@ -0,0 +1,330 @@
+# Copyright (c) 2021 elParaguayo
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# Widget specific tests
+
+import datetime
+import sys
+from importlib import reload
+from time import sleep
+
+import pytest
+
+import libqtile.config
+from libqtile.bar import Bar
+from libqtile.widget import clock
+
+
+def no_op(*args, **kwargs):
+    pass
+
+
+class FakeWindow:
+    class _NestedWindow:
+        wid = 10
+    window = _NestedWindow()
+
+
+class MockDatetime(datetime.datetime):
+
+    @classmethod
+    def now(cls, *args, **kwargs):
+        return cls(2021, 1, 1, 10, 20, 30)
+
+    def astimezone(self, tzone=None):
+        if tzone is None:
+            return self
+        else:
+            return self + datetime.timedelta(hours=tzone)
+
+
+def test_clock(fake_qtile, monkeypatch):
+    """ test clock output with default settings """
+    monkeypatch.setattr("libqtile.widget.clock.datetime", MockDatetime)
+    clk1 = clock.Clock()
+    fakebar = Bar([clk1], 24)
+    fakebar.window = FakeWindow()
+    fakebar.width = 10
+    fakebar.height = 10
+    fakebar.draw = no_op
+    clk1._configure(fake_qtile, fakebar)
+    text = clk1.poll()
+    assert text == "10:20"
+
+
+def test_clock_invalid_timezone(fake_qtile, monkeypatch):
+    """ test clock widget with invalid timezone (and no pytz or dateutil modules) """
+    class FakeDateutilTZ:
+        @classmethod
+        def tz(cls):
+            return cls
+
+        @classmethod
+        def gettz(cls, val):
+            return None
+
+    # Stop system importing these modules in case they exist on environment
+    monkeypatch.setitem(sys.modules, "pytz", None)
+    monkeypatch.setitem(sys.modules, "dateutil", None)
+    monkeypatch.setitem(sys.modules, "dateutil.tz", None)
```

Comment:
I suspect a lot of this patching could be put into a fixture as it's repeated in tests.

Reply:
Yeah. Why not do it?

I use `@pytest.mark.usefixtures(​...)` for this kind of fixture that doesn't need to pass an argument.

Reply:
Ah. Hadn't understood how `usefixtures` works. That will make things much neater/more readable. Will update.